test/e2e/network/dual_stack.go

```diff
@@ -765,6 +717,32 @@ func validateServiceAndClusterIPFamily(svc *v1.Service, expectedIPFamilies []v1.
 	}
 }
 
+func validateEndpointsAndEndpointSlices(ctx context.Context, f *framework.Framework, svc *v1.Service, expectedIPFamilies []v1.IPFamily) {
+	if err := wait.PollUntilContextTimeout(ctx, 500*time.Millisecond, 10*time.Second, true, func(ctx context.Context) (bool, error) {
+		endpoint, err := f.ClientSet.CoreV1().Endpoints(svc.Namespace).Get(ctx, svc.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+
+		validateEndpointsBelongToIPFamily(svc, endpoint, expectedIPFamilies[0] /* endpoint controller works on primary IP */)
+		return true, nil
+	}); err != nil {
+		framework.Failf("Get endpoints for service %s/%s failed (%s)", svc.Namespace, svc.Name, err)
+	}
+
+	if err := wait.PollUntilContextTimeout(ctx, 500*time.Millisecond, 10*time.Second, true, func(ctx context.Context) (bool, error) {
+		slices, err := f.ClientSet.DiscoveryV1().EndpointSlices(svc.Namespace).List(ctx, metav1.ListOptions{LabelSelector: discoveryv1.LabelServiceName + "=" + svc.Name})
+		if err != nil || len(slices.Items) == 0 {
+			return false, nil
+		}
+
+		validateEndpointSlicesBelongToIPFamilies(svc, slices.Items, expectedIPFamilies)
```

Comment:
if you call `framework.Failf` inside this loop it will panic, you can return an error and do something like

```suggestion
		return true, validateEndpointSlicesBelongToIPFamilies(svc, slices.Items, expectedIPFamilies)
```

Reply:
it seems the other validate thing does the same, but that will panic the test

Reply:
> if you call `framework.Failf` inside this loop it will panic

yeah... um... why? It's not in another goroutine?

Reply:
oh, wait, duh, it is.
sorry, was confused by the fact that the existing code already does that. I guess it just didn't normally fail so no one really noticed...